src/Common/src/TypeSystem/Interop/IL/NativeStructType.cs

```diff
@@ -0,0 +1,373 @@
+                    }
+                }
+
+                Debug.Assert(layoutIndex == layout.Offsets.Length);
+            }
+            else
+            {
+                result.Offsets = null;
+            }
+
+            return result;
+        }
+
+        public override bool HasCustomAttribute(string attributeNamespace, string attributeName)
+        {
+            return false;
+        }
+
+        public override IEnumerable<MetadataType> GetNestedTypes()
+        {
+            return Array.Empty<MetadataType>();
+        }
+
+        public override MetadataType GetNestedType(string name)
+        {
+            return null;
+        }
+
+        protected override MethodImplRecord[] ComputeVirtualMethodImplsForType()
+        {
+            return Array.Empty<MethodImplRecord>();
+        }
+
+        public override MethodImplRecord[] FindMethodsImplWithMatchingDeclName(string name)
+        {
+            return Array.Empty<MethodImplRecord>();
+        }
+
+        public override int GetHashCode()
+        {
+            var hashCodeBuilder = new Internal.NativeFormat.TypeHashingAlgorithms.HashCodeBuilder(Namespace);
+
+            if (Namespace.Length > 0)
+            {
+                hashCodeBuilder.Append(".");
+            }
+
+            hashCodeBuilder.Append(Name);
+            return hashCodeBuilder.ToHashCode();
+        }
+
+        protected override TypeFlags ComputeTypeFlags(TypeFlags mask)
+        {
+            TypeFlags flags = 0;
+            if ((mask & TypeFlags.ContainsGenericVariablesComputed) != 0)
+            {
+                flags |= TypeFlags.ContainsGenericVariablesComputed;
+            }
+
+            if ((mask & TypeFlags.HasGenericVarianceComputed) != 0)
+            {
+                flags |= TypeFlags.HasGenericVarianceComputed;
+            }
+
+            if ((mask & TypeFlags.CategoryMask) != 0)
+            {
+                flags |= TypeFlags.ValueType;
+            }
+
+            return flags;
+        }
+
+        public override IEnumerable<FieldDesc> GetFields()
+        {
+            return _fields;
+        }
+
+        /// <summary>
+        /// Synthetic field on <see cref="NativeStructType"/>.
+        /// </summary>
+        private class NativeStructField : FieldDesc
+        {
+            private FieldDesc _managedField;
+            private TypeDesc _fieldType;
+            private NativeStructType _owningType;
+
+            private string _name;
+
+            public override TypeSystemContext Context
+            {
+                get
+                {
+                    return _owningType.Context;
+                }
+            }
+
+            public override TypeDesc FieldType
+            {
+                get
+                {
+                    return _fieldType;
+                }
+            }
+
+            public override bool HasRva
+            {
+                get
+                {
+                    return _managedField.HasRva;
```

Comment:
For the comments below, I think in general we should only preserve the characteristics that are native-only concepts for interop and leave the rest. Otherwise marshaling could end up with surprising behavior. Threadstatic is a very good example.